packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart

```diff
@@ -0,0 +1,196 @@
+
+/// Signature of an asynchronous function for astablishing a JSON RPC-2
+/// connection to a [Uri].
+typedef Future<json_rpc.Peer> RpcPeerConnectionFunction(Uri uri);
+
+/// [DartVm] uses this function to connect to the Dart VM on Fuchsia.
+///
+/// This function can be assigned to a different one in the event that a
+/// custom connection function is needed.
+RpcPeerConnectionFunction fuchsiaVmServiceConnectionFunction = _waitAndConnect;
+
+/// Attempts to connect to a Dart VM service.
+///
+/// Gives up after `_kConnectTimeout` has elapsed.
+Future<json_rpc.Peer> _waitAndConnect(Uri uri) async {
+  final Stopwatch timer = new Stopwatch()..start();
+
+  Future<json_rpc.Peer> attemptConnection(Uri uri) async {
+    WebSocket socket;
+    json_rpc.Peer peer;
+    try {
+      socket = await WebSocket.connect(uri.toString());
+      peer = new json_rpc.Peer(new IOWebSocketChannel(socket).cast())..listen();
+      return peer;
+    } catch (e) {
+      await peer?.close();
+      await socket?.close();
+      if (timer.elapsed < _kConnectTimeout) {
+        _log.info('Attempting to reconnect');
+        await new Future<Null>.delayed(_kReconnectAttemptInterval);
+        return attemptConnection(uri);
+      } else {
+        _log.severe('Connection to Fuchsia\'s Dart VM timed out at '
+            '${uri.toString()}');
+        rethrow;
+      }
+    }
+  }
+
+  return attemptConnection(uri);
+}
+
+/// Restores the VM service connection function to the default implementation.
+void restoreVmServiceConnectionFunction() {
+  fuchsiaVmServiceConnectionFunction = _waitAndConnect;
+}
+
+/// An error raised when a malformed RPC response is received from the Dart VM.
+///
+/// A more detailed description of the error is found within the [message]
+/// field.
+class RpcFormatError extends Error {
+  /// The reason for format error.
+  final String message;
+
+  /// Basic constructor outlining the reason for the format error.
+  RpcFormatError(this.message);
+
+  @override
+  String toString() {
+    return '$RpcFormatError: $message\n${super.stackTrace}';
+  }
+}
+
+/// Handles JSON RPC-2 communication with a Dart VM service.
+///
+/// Either wraps existing RPC calls to the Dart VM service, or runs raw RPC
+/// function calls via [invokeRpc].
+class DartVm {
+  final json_rpc.Peer _peer;
+
+  DartVm._(this._peer);
+
+  /// Attempts to connect to the given [Uri].
+  ///
+  /// Throws an error if unable to connect.
+  static Future<DartVm> connect(Uri uri) async {
+    if (uri.scheme == 'http') {
+      uri = uri.replace(scheme: 'ws', path: '/ws');
+    }
+    final json_rpc.Peer peer = await fuchsiaVmServiceConnectionFunction(uri);
+    if (peer == null) {
+      return null;
+    }
+    return new DartVm._(peer);
+  }
+
+  /// Invokes a raw JSON RPC command with the VM service.
+  ///
+  /// When `timeout` is set and reached, throws a [TimeoutException].
+  ///
+  /// If the function returns, it is with a raw JSON response.
```

Comment:
a raw JSON response would be a string, this returns a map. Maybe "a parsed JSON object"?

Reply:
Fixed.